pymc_experimental/inference/laplace.py

```diff
@@ -0,0 +1,183 @@
+#   Copyright 2024 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from collections.abc import Sequence
+from typing import Optional
+
+import arviz as az
+import numpy as np
+import pymc as pm
+import xarray as xr
+from arviz import dict_to_dataset
+from pymc.backends.arviz import (
+    coords_and_dims_for_inferencedata,
+    find_constants,
+    find_observations,
+)
+from pymc.model.transform.conditioning import remove_value_transforms
+from pymc.util import RandomSeed
+from pytensor import Variable
+
+
+def laplace(
+    vars: Sequence[Variable],
+    draws=1_000,
+    chains=1,
+    model=None,
+    random_seed: Optional[RandomSeed] = None,
+    progressbar=True,
+):
+    """
+    Create a Laplace (quadratic) approximation for a posterior distribution.
+
+    This function generates a Laplace approximation for a given posterior distribution using a specified
+    number of draws. This is useful for obtaining a parametric approximation to the posterior distribution
+    that can be used for further analysis.
+
+    Parameters
+    ----------
+    vars : Sequence[Variable]
+        A sequence of variables for which the Laplace approximation of the posterior distribution
+        is to be created.
+    draws : int, optional, default=1_000
+        The number of draws to sample from the posterior distribution for creating the approximation.
+    chains : int, default=1
+        The number of chains to sample. For chains=0 only the fit of the Laplace approximation is
```

Comment:
Only returning the `fit` should be draws=0?

Re: chains, right now it doesn't make much sense because all chains will be the same. This changes if: 1) Users can pass different starts for different chains or 2) We add some random jitter across the initial point so that multiple chains could indeed arrive at different conclusions. I'm not sure this is useful, so maybe limiting to 1 chain / optimization (and removing the kwarg) may be fine?

Reply:
This was suggested by @theorashid

Reply:
Just adding that it is pedagogically confusing to have multiple chains, because the user might try to incorrectly use MCMC diagnostics on the posterior to see if their "Laplace approximation" has converged. This is (probably) the same reason that the variation inference objects return idata with only one chain (the user again cannot specify multiple chains).

Reply:
I think it makes sense to limit to 1 chain and not having chains as part of the method arguments.

Reply:
The latest commit fixes this issue.